src/derivatives.jl

```diff
@@ -0,0 +1,162 @@
+function derivative(A::LinearInterpolation{<:AbstractVector{<:Number}}, t::Number)
+    idx = findfirst(x -> x >= t, A.t) - 1
+    idx == 0 ? idx += 1 : nothing
+    θ = 1 / (A.t[idx+1] - A.t[idx])
+    (A.u[idx+1] - A.u[idx]) / (A.t[idx+1] - A.t[idx])
+end
+
+function derivative(A::LinearInterpolation{<:AbstractMatrix{<:Number}}, t::Number)
+    idx = findfirst(x -> x >= t, A.t) - 1
+    idx == 0 ? idx += 1 : nothing
+    θ = 1 / (A.t[idx+1] - A.t[idx])
+    (A.u[:, idx+1] - A.u[:, idx]) / (A.t[idx+1] - A.t[idx])
+end
+
+function derivative(A::QuadraticInterpolation{<:AbstractVector{<:Number}}, t::Number)
+    idx = findfirst(x -> x >= t, A.t) - 1
+    idx == 0 ? idx += 1 : nothing
+    if idx == length(A.t) - 1
+        i₀ = idx - 1; i₁ = idx; i₂ = i₁ + 1;
+    else
+        i₀ = idx; i₁ = i₀ + 1; i₂ = i₁ + 1;
+    end
+    dl₀ = (2t - A.t[i₁] - A.t[i₂]) / ((A.t[i₀] - A.t[i₁]) * (A.t[i₀] - A.t[i₂]))
+    dl₁ = (2t - A.t[i₀] - A.t[i₂]) / ((A.t[i₁] - A.t[i₀]) * (A.t[i₁] - A.t[i₂]))
+    dl₂ = (2t - A.t[i₀] - A.t[i₁]) / ((A.t[i₂] - A.t[i₀]) * (A.t[i₂] - A.t[i₁]))
+    A.u[i₀] * dl₀ + A.u[i₁] * dl₁ + A.u[i₂] * dl₂
+end
+
+function derivative(A::QuadraticInterpolation{<:AbstractMatrix{<:Number}}, t::Number)
+    idx = findfirst(x -> x >= t, A.t) - 1
+    idx == 0 ? idx += 1 : nothing
+    if idx == length(A.t) - 1
+        i₀ = idx - 1; i₁ = idx; i₂ = i₁ + 1;
+    else
+        i₀ = idx; i₁ = i₀ + 1; i₂ = i₁ + 1;
+    end
+    dl₀ = (2t - A.t[i₁] - A.t[i₂]) / ((A.t[i₀] - A.t[i₁]) * (A.t[i₀] - A.t[i₂]))
+    dl₁ = (2t - A.t[i₀] - A.t[i₂]) / ((A.t[i₁] - A.t[i₀]) * (A.t[i₁] - A.t[i₂]))
+    dl₂ = (2t - A.t[i₀] - A.t[i₁]) / ((A.t[i₂] - A.t[i₀]) * (A.t[i₂] - A.t[i₁]))
+    A.u[:, i₀] * dl₀ + A.u[:, i₁] * dl₁ + A.u[:, i₂] * dl₂
+end
+
+function derivative(A::LagrangeInterpolation{<:AbstractVector{<:Number}}, t::Number)
+    idxs = DataInterpolations.findRequiredIdxs(A, t)
+    if A.t[idxs[1]] == t
+        return zero(A.u[idxs[1]])
+    end
+    G = zero(A.u[1]); F = zero(A.t[1])
+    DG = zero(A.u[1]); DF = zero(A.t[1])
+    tmp = G
+    for i = 1:length(idxs)
+        mult = one(A.t[1])
+        for j = 1:(i - 1)
+            mult *= (A.t[idxs[i]] - A.t[idxs[j]])
+        end
+        for j = (i+1):length(idxs)
+            mult *= (A.t[idxs[i]] - A.t[idxs[j]])
+        end
+        wi = inv(mult)
```

Comment:
We probably want to cache the barycentric weights in the `LagrangeInterpolation` structure. Computing the weights takes O(n^2) time and is independent of evaluation. Evaluation takes only O(n) time. Not caching weights renders barycentric Lagrange interpolation as slow as the classical one.